### PR TITLE
Color scheme for spectrogram

### PR DESCRIPTION
**Pull request by dofuuz**
I made a post for this PR:
https://forum.audacityteam.org/viewtopic.php?f=20&t=117875

Same content, but with code for generating the colormap:
https://github.com/dofuuz/audacity-colormap

proposal

color-scheme

### CODE REVIEW

**Comment by Paul-Licameli** [comment]
We're still interested in this proposal.

Please see more discussion at the Audacity Forum.

**Comment by dofuuz** [comment]
@Paul-Licameli
Rebased commit at recent upstream master branch.

TBH, I'm not fully satisfied with CLA.
But I belive that commercial profit and support can boost imporvement on OSS.
So I signed at the CLA anyway.

I just wish Audacity would remain free, open and be **much greater** than now. 😀

File: src/prefs/SpectrogramSettings.h [hunk 1]
```diff
@@ -128,7 +129,17 @@ class AUDACITY_DLL_API SpectrogramSettings : public PrefsListener
    size_t GetFFTLength() const; // window size (times zero padding, if STFT)
    size_t NBins() const;
 
-   bool isGrayscale;
+   typedef int ColorScheme;
+   enum ColorSchemeValues : int {
+      // Keep in correspondence with AColor::gradient_pre
```
Comment (Paul-Licameli) [inline src/prefs/SpectrogramSettings.h:134]: You mean `AColor::colorSchemes`

File: src/prefs/SpectrogramSettings.cpp [hunk 2]
```diff
@@ -154,6 +154,23 @@ const EnumValueSymbols &SpectrogramSettings::GetScaleNames()
    return result;
 }
 
+//static
+const EnumValueSymbols &SpectrogramSettings::GetColorSchemeNames()
+{
+   static const EnumValueSymbols result{
+      // Keep in correspondence with enum SpectrogramSettings::ColorScheme:
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/prefs/SpectrogramSettings.cpp:161]: Also, make a non-static assertion about that correspondence, like
`
wxASSERT(result.size() == csNumColorScheme);
`

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]
I'll complete this first review for now, but I need to examine AColor.cpp more closely, and consider whether there might be another fix for the sync lock tiles.

**Comment by Paul-Licameli** [comment]
> @Paul-Licameli
> Rebased commit at recent upstream master branch.
> 
> TBH, I'm not fully satisfied with CLA.
> But I belive that commercial profit and support can boost imporvement on OSS.
> So I signed at the CLA anyway.
> 
> I just wish Audacity would remain free, open and be **much greater** than now. 😀

As for that, I share those hopes.

I not only hope so, I expect so, from what I know.

Though that is on the understanding that code of the Audacity audio editor "remaining free" is consistent with the possibility of a dual sub-licensing of parts of it that the CLA leaves open.

File: src/prefs/SpectrogramSettings.h [hunk 3]
```diff
@@ -128,7 +129,17 @@ class AUDACITY_DLL_API SpectrogramSettings : public PrefsListener
    size_t GetFFTLength() const; // window size (times zero padding, if STFT)
    size_t NBins() const;
 
-   bool isGrayscale;
+   typedef int ColorScheme;
+   enum ColorSchemeValues : int {
```
[the remark this answers is not in the document]
Comment (dofuuz) [reply src/prefs/SpectrogramSettings.h:133]: This change breaks template type deduction and cause compile errors. (At least at VS2017)
I won't apply this.

File: src/AColor.h [hunk 4]
```diff
@@ -123,8 +123,9 @@ class AUDACITY_DLL_API AColor {
    static wxBrush tooltipBrush;
 
    static bool gradient_inited;
-   static const int gradientSteps = 512;
-   static unsigned char gradient_pre[ColorGradientTotal][2][gradientSteps][3];
+   static const int colorSchemes = 4;
+   static const int gradientSteps = 256;
```
[the remark this answers is not in the document]
Comment (dofuuz) [reply src/AColor.h:127]: To save memory and gain some speed. It might be trivial.
Anyway, this does not affect visual quality.

File: src/prefs/SpectrogramSettings.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/prefs/SpectrogramSettings.h:133]: OK ignore that suggestion

File: src/prefs/SpectrogramSettings.cpp [hunk 5]
```diff
@@ -254,7 +272,7 @@ void SpectrogramSettings::LoadPrefs()
 
    gPrefs->Read(wxT("/Spectrum/WindowType"), &windowType, eWinFuncHann);
 
-   isGrayscale = (gPrefs->Read(wxT("/Spectrum/Grayscale"), 0L) != 0);
+   colorScheme = ColorScheme(gPrefs->Read(wxT("/Spectrum/ColorScheme"), 0L));
```
[the remark this answers is not in the document]
Comment (dofuuz) [reply src/prefs/SpectrogramSettings.cpp:275]: Spectrogram has per-track setting. Can it really be static? I'm confused.

File: src/AColor.cpp [hunk 6]
```diff
@@ -635,13 +635,23 @@ void AColor::DarkMIDIChannel(wxDC * dc, int channel /* 1 - 16 */ )
...
 
 
-unsigned char AColor::gradient_pre[ColorGradientTotal][2][gradientSteps][3];
+unsigned char AColor::gradient_pre[ColorGradientTotal][colorSchemes][gradientSteps][3];
 
 void AColor::PreComputeGradient() {
    if (gradient_inited) return;
    gradient_inited = 1;
 
+   // Keep in correspondence with enum SpectrogramSettings::ColorScheme
```
[the remark this answers is not in the document]
Comment (dofuuz) [reply src/AColor.cpp:644]: Added the static_assert in SpectrogramSettings.cpp, along with `wxASSERT(csNumColorScheme == result.size());`.
I coudln't find better place to put it.

File: src/prefs/SpectrogramSettings.h (hunk 1 above)
Comment (dofuuz) [reply src/prefs/SpectrogramSettings.h:134]: Both of them.

File: src/prefs/SpectrogramSettings.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dofuuz) [reply src/prefs/SpectrogramSettings.cpp:161]: Done

File: src/commands/SetTrackInfoCommand.h [hunk 7]
```diff
@@ -137,7 +136,6 @@ class SetTrackVisualsCommand : public SetTrackBase
 
    bool bHasUseSpecPrefs;
    bool bHasSpectralSelect;
-   bool bHasGrayScale;
```
[the remark this answers is not in the document]
Comment (dofuuz) [reply src/commands/SetTrackInfoCommand.h:140]: Added 'Spectro. Color scheme' to 'Set Track Visuals' macro command.

File: src/AColorResources.h [hunk 8]
```diff
@@ -1,3 +1,15 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  AColorResources.h
```
[the remark this answers is not in the document]
Comment (dofuuz) [reply src/AColorResources.h:5]: It's included from AColor.cpp.
It contains RGB value of new spectrogram colormap.

File: src/AColor.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/AColor.h:127]: ok leave it so

File: src/AColorResources.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/AColorResources.h]: Now I see, you added the #include in the first commit.  I thought it would be in the same commit that added AColorResources.h.

This means the first commit can't compile. I prefer that each commit in a commit history should compile.

After more complete review, I would like the new commits to be squashed, but it is also good to leave the first refactoring commit separated, and fixed so it compiles.

File: src/prefs/SpectrogramSettings.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/prefs/SpectrogramSettings.cpp:280]: The old code here is in fact using a global setting.

The system of spectrogram settings leaves something to be desired -- a track can have independent settings, or it can take its settings from defaults, but it can't default individual settings separately.

Don't try to understand and fix all of that -- just understand that where you rewrite usages of a global setting (which are what gPrefs contains), then it is preferable to use an EnumSetting object instead.

File: src/prefs/SpectrogramSettings.cpp [hunk 9]
```diff
@@ -254,7 +272,7 @@ void SpectrogramSettings::LoadPrefs()
 
    gPrefs->Read(wxT("/Spectrum/WindowType"), &windowType, eWinFuncHann);
 
-   isGrayscale = (gPrefs->Read(wxT("/Spectrum/Grayscale"), 0L) != 0);
```
[the remark this answers is not in the document]
Comment (dofuuz) [reply src/prefs/SpectrogramSettings.cpp:257]: Done.

File: src/prefs/SpectrogramSettings.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (dofuuz) [reply src/prefs/SpectrogramSettings.cpp]: Done.

File: src/AColor.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (dofuuz) [reply src/AColor.h:127]: Applied with preference migration 😉

**Comment by dofuuz** [comment]
@Paul-Licameli Ping

**Comment by Paul-Licameli** [comment]
I have not forgotten this!  I will write more today.

File: src/prefs/SpectrogramSettings.cpp [hunk 10]
```diff
@@ -160,14 +160,18 @@ const EnumValueSymbols &SpectrogramSettings::GetColorSchemeNames()
...
-      XO("Color (from Theme)") ,
+      { wxT("SpecColorTheme"), XO("Color (from Theme)") },
       /* i18n-hint: grayscale color scheme for spectrograms */
-      XO("Grayscale") ,
+      { wxT("SpecGrayscale"), XO("Grayscale") },
       /* i18n-hint: inverse grayscale color scheme for spectrograms */
-      XO("Inv. Grayscale") ,
+      { wxT("SpecInvGrayscale"), XO("Inv. Grayscale") },
    };
+
```
Comment (Paul-Licameli) [inline src/prefs/SpectrogramSettings.cpp:172]: Thank you for these changes.

But now, we are in the period of "string freeze."  That means the message catalog locale/audacity.pot, in which all the XO (and XC) strings are extracted, is in a final form, allowing translators to finish their work.  One of our designers reviewed this project, and specified changes to the user-visible strings.

So you must:

1) change the XO strings above to the XC strings below.  (But leave the wxT strings as they are.)
2) Find these strings in lib-strings/locale/UnusedStrings.h; be sure you have copied them exactly; and delete them from that file.
3) Rebase this branch onto more recent master.  (And while doing that, do an interactive rebase, and make sure all commits in the sequence compile.  I believe the first one does not.  And maybe do some fixups to shorten the list of commits.)
4) Test (using one of the completed languages, such as Dutch) that translation works.  (You change language using the Interface page of Preferences.)
5) Force-push your rebased branch.

'
XC("Color (default)",   "spectrum prefs"),
XC("Color (classic)",   "spectrum prefs"),
XC("Grayscale",         "spectrum prefs"),
XC("Inverse grayscale", "spectrum prefs"),
'

File: src/prefs/SpectrumPrefs.cpp [hunk 11]
```diff
@@ -223,11 +223,12 @@ void SpectrumPrefs::PopulateOrExchange(ShuttleGui & S)
                S.Id(ID_FREQUENCY_GAIN).TieNumericTextBox(XXO("High &boost (dB/dec):"),
                mTempSettings.frequencyGain,
                8);
+
+            S.Id(ID_COLOR_SCHEME).TieChoice(XXO("Color Sche&me:"),
```
Comment (Paul-Licameli) [inline src/prefs/SpectrumPrefs.cpp:227]: Also for string freeze, the string above must be exactly as the designer specified in UnusedStrings.h.  Cut and paste this string from there (with the comment too, pasted to the line before the line with the string):

'// i18n-hint Scheme refers to a color scheme for spectrogram colors
XC("Sche&me",            "spectrum prefs"),'

File: src/commands/SetTrackInfoCommand.cpp [hunk 12]
```diff
@@ -357,6 +360,14 @@ void SetTrackVisualsCommand::PopulateOrExchange(ShuttleGui & S)
       S.Optional( bHasSpectralSelect ).TieCheckBox( XXO("Spectral Select"),    bSpectralSelect);
    }
    S.EndMultiColumn();
+   S.StartMultiColumn(3, wxEXPAND);
+   {
+      S.SetStretchyCol( 2 );
+      auto schemes = SpectrogramSettings::GetColorSchemeNames();
+      S.Optional( bHasSpecColorScheme).TieChoice( XXO("Spectro. Color scheme:"), mSpecColorScheme,
```
Comment (Paul-Licameli) [inline src/commands/SetTrackInfoCommand.cpp:367]: Also for string freeze, it's too late now to add new user-visible strings -- or else they won't translate from English.  So change this to reuse the same string as in the preferences dialog:

`XC("Sche&me", "spectrum prefs")`

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]
Changes to user-visible strings because of string freeze

File: src/prefs/SpectrogramSettings.cpp (hunk 10 above)
Comment (dofuuz) [reply src/prefs/SpectrogramSettings.cpp:172]: Done

**Comment by petersampsonaudacity** [comment]
I just tested on W10 with dofuuz's latest build and I see that we now have the changed color descriptions in Spectrogram Preferences as we discussed.  Thanks for sorting that.
image

**Comment by petersampsonaudacity** [comment]
I tested with pure Spectrogram and with Multi-view dor the new and alternative colorways

**Comment by Paul-Licameli** [comment]
Thanks, Peter.  Did you also observe the changes in Set Track Visuals?  A choice replaces what was only a checkbox for grayscale.

**Comment by petersampsonaudacity** [comment]
Paul, I assume you mean from the TCp - in which case yes, I have been playing with that too and all looks fine.

**Comment by Paul-Licameli** [comment]
> Paul, I assume you mean from the TCp - in which case yes, I have been playing with that too and all looks fine.

No, I refer to one of the Macro commands.  See #969

**Comment by Paul-Licameli** [comment]
I'm pleased with these changes to address my points.  Each commit builds.

I am satisfied too now with changes in AColor.cpp.  (I might suggest that the names in AColorResources.h could be made static, to have no linkage and spare a little space in symbol tables.)

But I do not want the change in Experimental.h.  It leaves no visual indication of sync lock in a spectrogram.  If you think this change is very important, than I think some alternative solution should be sought.  The sync lock tiles might be painted more than once (as they are in Waveform view).  They might be repainted last, with transparency.  Perhaps you fill follow up this implementation suggestion.

So to summarize:   please rebase onto most master, and omit the change to Experimental.cmake.  Then I will approve it and merge it!

**Comment by dofuuz** [comment]
Removing transparency from spectrogram is important fix! Without the change, new colormap gonna lose its power.
I'll find if there are easy fix for this situation...

**Comment by petersampsonaudacity** [comment]
> But I do not want the change in Experimental.h. It leaves no visual indication of sync lock in a spectrogram. If you think this change is very important, than I think some alternative solution should be sought. The sync lock tiles might be painted more than once (as they are in Waveform view). They might be repainted last, with transparency. Perhaps you fill follow up this implementation suggestion.


But even in current Audacity there is very poor visual indication of sync-lock in a spectrogram ...

The example image below was made with 3.0.2
image

I thought we had a Bugzilla entry for this, but I can't find it.

**Comment by petersampsonaudacity** [comment]
>I thought we had a Bugzilla entry for this, but I can't find it. Maybe we should be logging one ...

I logged this as P3 Bug #2806
https://bugzilla.audacityteam.org/show_bug.cgi?id=2806
Spectrogram view has poor/inadequate visual cues to show Sync-lock selection

And remember that this is not a result of the new Spectrogram colorway - it's always been like this.

**Comment by Paul-Licameli** [comment]
> Removing transparency from spectrogram is important fix! Without the change, new colormap gonna lose its power.
> I'll find if there are easy and nice fix for this situation...

As your branch is now, there is no indication at all of sync lock, rather than the existing one, poor though that might be.

Why do you think removing transparency of the spectrogram is very important?  In my own trials building your branch,  I did not think it looked bad either with or without it.

Would you consider merging the branch with transparency, but opening an issue to correct the problems?

I see your example in "Obstacles." https://forum.audacityteam.org/viewtopic.php?f=20&t=117875#p424607

**Comment by SteveDaulton** [comment]
I don't think that losing the Sync-Lock "clocks" in spectrogram view is a problem. Even with "classic" spectrogram view they are barely visible with real world audio, and imo they are little more than visual clutter when they are. Waveform view, or multi-view are far better suited to temporal editing. I'd rather have a clearer view of the spectrogram in track spectrogram view than overlaid information that is irrelevant to "spectral" editing.

**Comment by dofuuz** [comment]
synclock
↑ New color with `SPECTROGRAM_OVERLAY` transparency

Transparency makes the visual totally different from what I intended(accurate and pretty).
Background becomes brighter than some of foreground. There are some grain around black which should not appear. It means loss of accuracy.
Overall color becomes desaturated. Not pretty.

Bringing clock overlay on top of spectrogram [like this](https://user-images.githubusercontent.com/8174871/122054382-96d9d400-ce22-11eb-9890-a7db4d253904.jpg) might be the fix. 
But basically I have same opinion with @SteveDaulton.

**Comment by Paul-Licameli** [comment]
If we want to keep the clocks but also make spectrogram non-transparent, then I think a solution is to draw the clocks with transparency, not the track, and paint the clocks later.

If we lose all of the clocks in spectral view, arguably that's a regression.  But maybe Steve and Peter would agree not to treat as such.

Are there examples of clocks on the spectrogram in the manual?  If so they would be one more thing to change in manual images.

I think the objection that clocks are clutter for spectral editing is irrelevant.  The track that you do interact with, for spectral editing, never shows the clocks.  The clocks indicate what part of another track will be affected, if the track you do interact with changes its length with insertion or deletion.  Spectral editing doesn't do that, but other edits do.

**Comment by Paul-Licameli** [comment]
> synclock
> ↑ New color with `SPECTROGRAM_OVERLAY` transparency. Clock background on 4th track is not visible well.
> 
> Transparency makes the visual totally different from what I intended(accurate and pretty).
> Background becomes brighter than some of foreground. There are some grain around black which should not appear. It means loss of accuracy.
> Overall color becomes desaturated. Not pretty.
> 
> Bringing clock overlay on top of spectrogram [like this](https://user-images.githubusercontent.com/8174871/122054382-96d9d400-ce22-11eb-9890-a7db4d253904.jpg) might be the fix.
> But basically I have same opinion with @SteveDaulton.

How did you do that "like this" picture?  Did you change Audacity source code to do it?

**Comment by dofuuz** [comment]
> How did you do that "like this" picture? Did you change Audacity source code to do it?
No. I used image editor.

**Comment by petersampsonaudacity** [comment]
> If we lose all of the clocks in spectral view, arguably that's a regression. But maybe Steve and Peter would agree not to treat it as such.

It's not really our call any more - But I would certainly be happy with the way it si now with the closcks invisible on the spectrogram (I agree with Steve's views above that  
a) they've been largely unvisible on Soectrograms anyway and 
b) other views are better for temporal editing)

So if MUSE QA and RM agree (Dmitry, Martin, Jouni) I would vote for this - and this would then let me close P1 Bug 2805.

As far as I can recall no user has ever complained about the clocks on Spectrograms - porr visibiliy of or lack thereof.

**Comment by Paul-Licameli** [comment]
I would leave bug 2806 open.  But we need to get a release out very soon, and these new colors will be a banner feature, and I don't want this detail to block it.

Maybe Myungchul will even write the fix easily.  I outlined how we might to do it.  Repaint the clocks, after the spectrum, making the clocks the transparency.

**Comment by dofuuz** [comment]
image
↑ Left: Light theme, Right: Dark theme

Fortunately, It seems like the fix is not hard to implement.
I'll push commit after some cleanup in couple of days. For now, I have to sleep😴

To adjust color and transparency of clock overlay, we need to edit theme image. I won't do this.
I think the clock should be gray, 50% transparent.

**Comment by SteveDaulton** [comment]
Thanks for trying the clocks dofuuz, but from a user perspective I would much prefer to NOT have the clocks.

When using track spectrogram view and Sync-Lock, I would have to temporarily turn off Sync-Lock to remove the distraction of the clocks. That's more work for the user, and a needless distraction from their workflow. The user can already see if Sync-Lock is on by looking for the clock in the control panel on the left hand end of the track - which is how users have had to work (with zero complaints) when using the "classic" spectrogram color scheme.

With QA hat on, usability trumps consistency. Now that I can see what it looks like with the new color scheme + clocks (thank you), I would say that it is significantly worse than not having the clocks showing in spectrograms.

Steve

**Comment by Paul-Licameli** [comment]
Thank you, Steve.

You have a strong opinion about the clocks.  I do NOT have a strong opinion about them -- which is why I brought the matter up:  in the absence of strong opinions, the presumption should be conservatism about the application's behavior, changing only what one needs to change for the really new feature.

In this case, hiding the clocks was not the intention of the feature, but a side-effect of trying to achieve something else, the non-transparency that improved spectrogram appearance.

However, given your remarks now,  I find no reason left to delay the merger of this pull request.  If someone really wants to overrule the elimination of the clocks, let them raise an issue.

**Review by Paul-Licameli — APPROVED** [review]

**Comment by Paul-Licameli** [comment]
Merged!

The residual issue about the clocks -- if anyone insists on keeping them -- is already at: https://bugzilla.audacityteam.org/show_bug.cgi?id=2806

**Comment by dofuuz** [comment]
Great😆
We can implement sync-lock overlay later, if there are anyone complain about it.

**Comment by Paul-Licameli** [comment]
dofuuz, do you wish to be listed in the Credits dialog of the next version?  If so, how exactly should we spell your name?  (In Hangul too if you like.)

**Comment by dofuuz** [comment]
@Paul-Licameli 
Yes. List me as
Myungchul Keum
Thanks😆

**Comment by petersampsonaudacity** [comment]
I have updated the Website Credits page adding Keum to the "Contributors"
https://www.audacityteam.org/about/credits/

I cannot update the in-app credits.

Thanks for your work on this Keum/dofuuz   😆

**Comment by Paul-Licameli** [comment]
Please help, we want to be sure that it is correct to alphabetize that under K and not M!

**Comment by dofuuz** [comment]
Keum is my family name.

**Comment by petersampsonaudacity** [comment]
Thanks for that clarification dofuuz ( Myungchul)  - I have adjusted that ordering in the Website credits:
https://www.audacityteam.org/about/credits/

Th latest 3/.0.3 alpha already has the correct alphabetic ordering.

Peter.
